### DISCUSSION TITLE
Why is "<:" used in the spec?

### DISCUSSION THREAD

**Discussion by Persola** [syntax-tree/unist#49] [Q&A] [answered]
I believe unist is using "<:" to indicate inheritance, but [the WebIDL spec](https://www.w3.org/TR/WebIDL/#dfn-inherit) just asks for ":". Am I missing something?

**Comment by wooorm** [comment, marked as answer]
I believe it was used in the original spidermonkey estree ast, but unfortunately that link just started 404ing: https://developer.mozilla.org/en-US/docs/SpiderMonkey/Parser_API.

**Reply by Persola** [reply]
👍

**Reply by burdiyan** [reply]
I was actually looking for a right place to ask a question about IDL so I guess I actually do it here.

I just started digging into unist, so I’m total noob, but I’m looking into defining a new AST for a custom format.

I see that unist spec and all the ASTs use very similar documentation format and this IDL which claims to be WebIDL-like but I guess it’s actually totally incompatible.

So I wonder:

1. Is the syntax of this IDL documented anywhere?
2. Are there any parsers for the IDL? 
3. Are there any code generation tools available, so that I could generate the actual code instead of duplicating the effort of writing and maintaining docs and code in sync.
4. Are these README files with different ASTs actually hand-written or generated somehow?

**Reply by wooorm** [reply]
The docs specifically mention “This specification is written in a Web IDL-like grammar”

1\. No, but it’s close to what the Spidermonkey AST had and what Web IDL is
2\. and 3. No. But if you use actual Web IDL, then yes.
4\. Handwritten

**Reply by burdiyan** [reply]
@wooorm Thanks for the prompt response!

Do you have any plans for creating tools for code generation (or the other way around, generating docs from code) so that documentation is up to date with the code, and/or adopting any existing general-purpose IDL be it Web-IDL, or even TypeScript definitions?

**Reply by ChristianMurphy** [reply]
I'm not aware of any IDL specific tools planned.
There was some discussion of JSDoc generating type definitions and readme documentation https://github.com/unifiedjs/rfcs/pull/5
But that's more at a utility/plugin level, than an AST definition level.

**Reply by wooorm** [reply]
@burdiyan Why are you wondering? If you want to create your own unist based format, you could choose how to write it. E.g., [`ntast` uses TS I believe](https://github.com/phuctm97/ntast).

**Reply by burdiyan** [reply]
@wooorm I'm actually looking into consuming the syntax tree from another language (Go in my case), so I was wondering if there were any language-agnostic tools available, that could help me save some time keeping JS, Go and docs in sync.

**Reply by ChristianMurphy** [reply]
It depends on what you mean by "consuming".
If you mean sending an AST from one to another, the ASTs are generally JSON, and most languages have a JSON parser/stringifier implementation.
If you mean parsing in different languages, check out https://www.antlr.org
If you mean validation, since ASTs are generally JSON, JSON schema is a fairly portable validation format.

**Reply by wooorm** [reply]
Also: we do have types for unist/mdast/hast in definitelytyped. If you want the packages, they're published at @types/mdast and such!

**Reply by burdiyan** [reply]
I'm still trying to figure out all the requirements between interop across languages. This discussion helped to clarify things. Thank you!
